define compound packets

kazuho opened the pull request:
In https://github.com/quicwg/base-drafts/issues/630#issuecomment-372985171 @janaiyengar wrote:
> Discussed at editors' meeting. We seemed to agree on something like option 4, with some specific rules: create "compound packet types" which have a length, and encode HS+0RTT and HS+1RTT long header packet types. To keep packet processing simple, the long header can be repeated so that a receiver can process parts of the packet under different keys in a straightforward manner.
> 
> This doesn't have to be the final resolution, but we should drive this issue to an end.
> 
> @kazuho : Would you be interested in driving this? Would you be able to write up a PR with a proposed design (use the one above if you like that design) and propose to the wg?

Thank you for your patience. This is the PR that does what is being suggested. Please take a look.

Some notes:

* a Compound-Initial packet is sent by client
* a Compound-Handshake packet is sent by both endpoints
  * a client can use it to send 0-RTT data along with packets ACKing the server's handshake packets prior to ServerFinished
  * a server can use it to send 0.5-RTT data along with ServerFinished
  * a client can use it to send 1-RTT data along with ClientFinished

* an endpoint is required to process the handshake portion of a compound packet even if the protected part is broken
  * the rationale behind is to not give an on-path attacker the chance of terminating the handshake simply by appending several bytes to the packet (currently, we require an on-path attacker to do AES based on the handshake protection key)

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -364,6 +364,8 @@ keys are established.
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|     [Payload Length (16)]     |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 |                          Payload (*)                        ...
```
Totally editorial:  You could put Payload in the second half of the previous line to avoid having a jagged illustration.  Variable-length fields like Payload don't have to have any particular width.

kazuho replied:
Thank you very much for the review. I've fixed this in a52715a. I will address other issues later (hopefully tomorrow).

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +687,34 @@ packet protection in detail.  After decryption, the plaintext consists of a
 
+## Compound Packets {#packet-compound}
+
+Compound packets allow the sender to assemble an Initial ({{packet-initial}})
```
I might use a different verb than "assemble"; perhaps "combine"?

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +687,34 @@ packet protection in detail.  After decryption, the plaintext consists of a
+or a Handshake ({{packet-handshake}}) packet and a Protected
+({{packet-protected}}) packet into one UDP packet, thereby reducing the number
+of packets required to be emitted when application data can be sent during the
+handshake.
```
Question, and important to discuss either way:  When sending Compound Initial + 0-RTT, does the minimum packet size apply to the entire compound and not just the Initial section?  If so, might be worth mentioning that one intended use-case of this is to meet the minimum with 0-RTT data instead of PADDING.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +687,34 @@ packet protection in detail.  After decryption, the plaintext consists of a
+exception being that it has the Payload Length field.  The field designates the
+length of the payload of the Handshake packet.  The remainder of the UDP packet
+contains either a 0-RTT Protected packet or a short header packet
+({{short-header}}).
```
It might be worth explaining or giving an example that all the packet header fields are there for the subsequent packet.  If that includes the type, by the way, you probably don't actually have to specify what type will be there, since the packet will declare it.

MikeBishop commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +687,34 @@ packet protection in detail.  After decryption, the plaintext consists of a
+The receiver of a Compound packet MUST individually process the Cryptographic
+Handshake packet and the Protected packet being contained.  It MUST process the
+Cryptographic Handshake packet of a Compound packet even when the Protected
+packet being contained is deemed invalid.
```
I think I'd simplify away these terms for the sections of the packet.  Simply say that the thing which follows is processed as a separate packet, which might be discarded, buffered, or processed depending on the receiver's current state.

MikeBishop replied:
...and contains its own packet number, and therefore gets acknowledged separately.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +685,34 @@ packet protection in detail.  After decryption, the plaintext consists of a
+A Compound-Initial packet is identical to an Initial packet with the exception
+being that it has the Payload Length field.  The field designates the length of
+the payload of the Initial packet.  The remainder of the UDP packet contains a
+0-RTT Protected packet.
```
What if the remainder of the packet doesn't contain a 0-RTT packet?

kazuho commented:
@MikeBishop @martinthomson Thank you for your comments. I think I have addressed all of them. Please take a look.

And I think that the text has become clearer by saying that a compound _UDP_ packet is something that consists of a *-with-Length packet and a protected packet. Thank you to @martinthomson for the suggestion.

ekr commented:
This seems OK, but why bother to have extra packet types. We've already expanded the long header quite a bit to make room for the two directional CIDs. Why not just add a length varint and have only one header form

ianswett commented:
In regards to @ekr.  If we're going to make the length varint and always present(which I'm happy with), can we make the packet number varint(but not relative) too for consistency?

mikkelfj commented:
Note that varint packet numbers has on effect on my proposed variation of #1079 PNE because it is immediately obvious where the PN begins if you encrypt and attach to the end. This is not an argument against, but a note to consider exposing length info elsewhere +/- privacy concerns.

ianswett commented:
For the context of this PR, I'd slightly favor adding a varint length to all long headers, as EKR suggested.

We can discuss packet numbers elsewhere, because as you said @mikkelfj, it may be complicated.

mikkelfj commented:
If we assume packet numbers are vartiable length and the PNE scheme is modified to attach the PN at the end as I proposed earlier, then the following extension could handle the length encoding AND provide a fast sanity check of packet validity.

Given the cleartext PN in a 32-bit word (assuming we never encode more than that), the length is given as a 2 bit number. This length can be stored in the last octet of packet, after the encrypted PN. The remaining 6 bits are used as follows:

C std lib has fast low quality PRNG:

int rand() {
  return (((holdrand = holdrand * 214013L + 2531011L) >> 16) & 0x7fff);
}

PN_checksum = (((PN * 214013L + 2531011L) >> 22) & 0xfc) | (PN_len - 1)
where PN_len is length in 1 to 4 octets.

The exact details may vary, but the point is a single fast 32-bit multiplication.

The length of the packet number can now be immediately decoded from the tail byte.
The packet number checksum can be verified immediately after decrypting the packet number.
If it fails there is roughly (and biased) a 1 / 64 chance that a packet with random content will have matching PN checksum. This can weed out trial decryption of the entire packet very fast. This in turn, can be used as a fast packet filter - even without having the full traffic key but just the less sensitive PN key.

Note that exposing the PN length can contribute to connection linkage since it says something about the connection data volume or age.

kazuho commented:
@MikeBishop @ianswett @ekr @mikkelfj Thank you all.

> Why not just add a length varint and have only one header form?

I will do this.

OTOH, I agree with @ianswett that how PN of a long header is encoded should better be discussed separately. Reasons: i) they *can* be discussed separately ii) representation of a PN depends on #1079 iii) representation of a PN may depend on the outcome of the stream 0 redesign.

mikkelfj commented:
Oh sorry, I meant to post this PN scheme elsewhere - completely ignoring Ians request by mistake.

martinthomson commented:
Yes, let's solve one problem at a time.  The idea of a varint packet number is pretty appealing to me, and I don't *think* that the stream 0 changes will necessarily change it, but we should probably wait and see.

martinthomson commented on draft-ietf-quic-tls.md:
```diff
@@ -552,7 +552,7 @@ older than 1.3 is negotiated.
 QUIC requires that the initial handshake packet from a client fit within the
 payload of a single packet.  The size limits on QUIC packets mean that a record
-containing a ClientHello needs to fit within 1171 octets.
+containing a ClientHello needs to fit within 1170 octets.
```
An 1170 octet ClientHello will have a two octet length, so this needs to be 1169.

kazuho replied:
At the moment the PR states that Payload Length field is a varint, and therefore the minimum overhead is 1 octet. Do you think that we should change it to a fix length of 2 octets?

martinthomson replied:
If the packet is this big, then the length will be 2 octets.  You can't have encode 1170 (or 1169) in one octet.  Obviously, if the ClientHello is smaller, then the length might only be 1 octet, but I think that one octet is impossible.  ClientHello.random is 32 octets and either a shares is at least 32 octets or a PSK binder is that big too.

kazuho replied:
Ah! Thank you for very much for the clarification.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -403,10 +405,13 @@ Packet Number:
 
+Payload Length:
+
+: The length of the Payload field in octets.
```
...encoded as a variable-length integer.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +691,22 @@ packet protection in detail.  After decryption, the plaintext consists of a
+## Compound Packets {#packet-compound}
+
+A sender can combine multiple QUIC packets (typically a Cryptographic Handshake
+packet and a Protected packet) into one UDP packet, thereby reducing the number
```
s/UDP packet/UDP datagram/

kazuho replied:
Done in 4504fdc.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +691,22 @@ packet protection in detail.  After decryption, the plaintext consists of a
+
+A sender can combine multiple QUIC packets (typically a Cryptographic Handshake
+packet and a Protected packet) into one UDP packet, thereby reducing the number
+of UDP packets required to be emitted when application data can be sent during
```
End the sentence at the comma.  Then "This can reduce the number of UDP datagrams that are needed to complete the handshake when multiple types of packets are sent at the same time."

kazuho replied:
Done in 5ed67ea.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +691,22 @@ packet protection in detail.  After decryption, the plaintext consists of a
+A sender can combine multiple QUIC packets (typically a Cryptographic Handshake
+packet and a Protected packet) into one UDP packet, thereby reducing the number
+of UDP packets required to be emitted when application data can be sent during
+the handshake.
```
Add: "A packet with a short header doesn't include a length, so it has to be the last packet included in a UDP datagram."

kazuho replied:
Done in 9dd629e.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -586,12 +596,15 @@ The first Initial packet that is sent by a client contains a randomized packet
+always starts at an offset of 0 (see {{stateless-retry}}) and the complete
+cryptographic handshake message MUST fit in a single packet (see {{handshake}}).
+
+The payload of a UDP packet carrying the Initial packet MUST be expanded to at
```
UDP datagram.

kazuho replied:
Done in 4504fdc.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -586,12 +596,15 @@ The first Initial packet that is sent by a client contains a randomized packet
+The payload of a UDP packet carrying the Initial packet MUST be expanded to at
+least 1200 octets (see {{packetization}}), by adding PADDING frames to the
+Initial packet and/or by combining the Initial packet with a Protected packet
+(see {{packet-compound}}).
```
s/Protected packet/0-RTT packet/

No point in being any less precise than needed.

kazuho replied:
Done in d5b7312.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +691,22 @@ packet protection in detail.  After decryption, the plaintext consists of a
+the handshake.
+
+The sender MUST NOT combine QUIC packets belonging to different QUIC
+connections into a single UDP packet.
```
datagram

kazuho replied:
Done in 4504fdc.

martinthomson commented on draft-ietf-quic-transport.md:
```diff
@@ -678,6 +691,22 @@ packet protection in detail.  After decryption, the plaintext consists of a
+Every QUIC packet that is being conveyed in a compound UDP packet is a complete
+QUIC packet.  No fields in the packet header are omitted.  The receiver of a
+compound UDP packet MUST individually process each QUIC packet and separately
+acknowledge them, as if they were received as distinct UDP packets.
```
s/as distinct UDP packets/as the payload of different UDP datagrams/

kazuho replied:
Done in 49d6f96.

kazuho commented:
@martinthomson Thank you for the super quick feedback. I think that I have addressed all your comments (as well as merging master) with the exception of https://github.com/quicwg/base-drafts/pull/1262#discussion_r180298854. Please take a look.

kazuho commented:
I believe that all known issues have been fixed in commits up to 93fe0be.

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -674,6 +689,24 @@ packet protection in detail.  After decryption, the plaintext consists of a
+
+The sender MUST NOT combine QUIC packets belonging to different QUIC
+connections into a single UDP datagram.
+
```
What about packets in the gray zone such as version negotiation and resets?

martinthomson replied:
Yes, I will address those in a follow-up.

mikkelfj commented on draft-ietf-quic-transport.md:
```diff
@@ -338,9 +340,14 @@ Packet Number:
+
+: The length of the Payload field in octets, encoded as a variable-length
+  integer ({{integer-encoding}}).
+
```
I came to think a bit about the variable payload length:

It makes it difficult to build a packet when you don't know the length in advance. Of course you can always just start with enough space. The intial packet starts with 1200 bytes (not sure about 0-RTT), and I don't think it can be more than 64K? So wouldn't a fixed 2 byte length field make more sense here?

martinthomson replied:
The variable length is primarily for consistency.  There are a few cases where packets will be small enough to benefit from the 6-bit length (Finished primarily), but the main benefit is that this is the same code throughout.  Better if we can manage to make the packet number encoding variable length as well (but that interacts in interesting ways with #1079, so that's not as clear a win as this).

janaiyengar commented:
Coming in late, but I like the final shape of this. Thanks, @kazuho, for driving this quickly!
